### PR TITLE
Change to LoggingEventBuilder

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

- Updates the static key string to a static start JSON string which reduces the number of StringBuilder.append calls
- `context` is tracked as null initially instead of `Collections.emptyList()`
- `addKeyValue(String, Object)` no longer eagerly converts the object to string and defers that until log time
- Split `performLogging` into two methods, one which has message formatting and another overload with no message formatting to remove a useless call to `MessageFormatter.arrayFormat(format, args)`
- Many single character String appends replaced with char appends
- Simplified type checking on `Object` to determine if it needs to be quoted
- Added simple check to determine when to add a `,` to remove a `deleteCharAt` call 

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/util/logging/LoggingEventBuilder.java [hunk 1]
```diff
@@ -126,7 +124,9 @@ public LoggingEventBuilder addKeyValue(String key, String value) {
      */
     public LoggingEventBuilder addKeyValue(String key, Object value) {
         if (this.isEnabled) {
-            addKeyValueInternal(key, value == null ? null : value.toString());
+            // Previously this eagerly called toString() on the value, but that can be expensive and unnecessary.
```
Comment (lmolkova) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/logging/LoggingEventBuilder.java:127]: no objections on doing `toString()` later, but the comment is not accurate - this is behind `isEnabled`, which is populated based on the log level

File: sdk/core/azure-core/src/main/java/com/azure/core/util/logging/LoggingEventBuilder.java [hunk 2]
```diff
@@ -250,7 +251,7 @@ public Throwable log(Throwable throwable) {
         Objects.requireNonNull(throwable, "'throwable' cannot be null.");
 
         if (this.isEnabled) {
-            performLogging(level, null, throwable);
+            performLogging(level, getMessageWithContext(null, throwable), logger.isDebugEnabled() ? throwable : null);
```
Comment (lmolkova) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/logging/LoggingEventBuilder.java:254]: same - what to do with an exception object is consumer, not producer decision

File: sdk/core/azure-core/src/main/java/com/azure/core/util/logging/LoggingEventBuilder.java [hunk 3]
```diff
@@ -386,49 +392,45 @@ static String writeJsonFragment(Map<String, Object> context) {
         }
 
         StringBuilder formatter = new StringBuilder(context.size() * 20);
+
+        // Keep track of whether we've written a value yet so we don't write a trailing comma.
+        // The previous implementation would delete the trailing comma, but internally this causes StringBuilder to
+        // copy the entirety of the string to a new buffer, which is very expensive.
```
Comment (lmolkova) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/logging/LoggingEventBuilder.java:398]: great catch!

File: sdk/core/azure-core/src/main/java/com/azure/core/util/logging/LoggingEventBuilder.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/logging/LoggingEventBuilder.java:127]: True that this is behind the enable flag but if someone forgets to actually call log this does waste cycles.

File: sdk/core/azure-core/src/main/java/com/azure/core/util/logging/LoggingEventBuilder.java [hunk 4]
```diff
@@ -222,7 +222,8 @@ public void log(Supplier<String> messageSupplier) {
     public void log(Supplier<String> messageSupplier, Throwable throwable) {
         if (this.isEnabled) {
             String message = messageSupplier != null ? messageSupplier.get() : null;
-            performLogging(level, message, throwable);
+            performLogging(level, getMessageWithContext(message, throwable),
+                logger.isDebugEnabled() ? throwable : null);
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/logging/LoggingEventBuilder.java:226]: In main right now if debug logging is not enabled there isn't a null passed.

It calls through this https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core/src/main/java/com/azure/core/util/logging/LoggingEventBuilder.java#L351 which on line 346 you can see the call to `removeThrowable`. We then log with `tuple.getThrowable()` which will be null.

File: sdk/core/azure-core/src/main/java/com/azure/core/util/logging/LoggingEventBuilder.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/logging/LoggingEventBuilder.java:254]: Same comment as the other case

**Review by lmolkova — APPROVED** [review]
Thanks!

**Pull request merged**